### PR TITLE
Docker release automation

### PR DESCRIPTION
**Pull request by VoR0220**
This makes automated dockerfile deployment a reality. Also reorganizes the CI with better environmental names, more sensible organization in terms of the travis.yml, sensible installation dependencies (not installing things that are not solidity related unless they are for the purpose of the CI), and the creation of a build script for solidity and updating the docs to reflect its existence.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -143,42 +168,36 @@ cache:
...
-        - TRAVIS_RELEASE=On
-        - TRAVIS_TESTS=On
+    - test $SOLC_DOCS != On || ./scripts/docs.sh
+    - test $SOLC_TESTS != On || (cd $TRAVIS_BUILD_DIR && ./scripts/tests.sh )
 
+# This is the deploy target for the dockerfile. If we are pushing into a develop branch, it will be tagged
+# as a nightly and appended the commit of the branch it was pushed in. If we are pushing to master it will
+# be tagged as "stable" and given the version tag as well.
+after_success:
+    - if [ "$TRAVIS_BRANCH" == "master" ]; then
```
Comment (chriseth) [inline .travis.yml]: Can you rather check for the branch starting with `v`? Then it will be a tagged release. Oh and please turn this into a script and put it in the scripts directory.

File: .travis.yml [hunk 2]
```diff
@@ -143,42 +168,36 @@ cache:
...
+        docker login -u="$DOCKER_USERNAME" -p="$DOCKER_PASSWORD";
+        docker tag ethereum/solc:$TRAVIS_COMMIT ethereum/solc:nightly;
+        docker push ethereum/solc:$TRAVIS_COMMIT;
+        docker push ethereum/solc:nightly;
+      elif [ "$TRAVIS_BRANCH" == "develop" ]; then
```
Comment (chriseth) [inline .travis.yml]: I think the two cases have to be swapped.

File: .travis.yml [hunk 3]
```diff
@@ -143,42 +168,36 @@ cache:
...
+        docker login -u="$DOCKER_USERNAME" -p="$DOCKER_PASSWORD";
+        docker tag ethereum/solc:$TRAVIS_COMMIT ethereum/solc:stable;
+        version=`grep -oP "PROJECT_VERSION \"?\K[0-9.]+(?=\")"? CMakeLists.txt`;
```
Comment (chriseth) [inline .travis.yml]: I think if we can directly use the branch name here.

File: scripts/Dockerfile [hunk 4]
```diff
@@ -1,12 +1,19 @@
...
+#Determine whether or not we are in the testing suite
+ARG TEST=Off
+#Copy working directory on travis to the image
+COPY / $WORKDIR
+
+#Install dependencies, eliminate annoying warnings, test if we are in a test environment, if so, 
+#add in soltest to the binary. If not, assume this is basic solidity image, build it and ship it out.
+RUN ./scripts/install_deps.sh && sed -i -E -e 's/include <sys\/poll.h>/include <poll.h>/' /usr/include/boost/asio/detail/socket_types.hpp			&&\
+test $TEST == On && cmake -DCMAKE_BUILD_TYPE=Release -DSTATIC_LINKING=1 || cmake -DCMAKE_BUILD_TYPE=Release -DTESTS=0 -DSTATIC_LINKING=1			&&\
+test $TEST == On && (make && install -s solc/solc /usr/bin && install -s test/soltest /usr/bin) || (make solc && install -s  solc/solc /usr/bin)	&&\
```
Comment (chriseth) [inline scripts/Dockerfile]: I don't think it makes sense to run the tests like that. In any case, please move that to its own PR.

File: docs/installing-solidity.rst [hunk 5]
```diff
@@ -197,7 +197,13 @@ Building Solidity is quite similar on Linux, macOS and other Unices:
     cd build
     cmake .. && make
 
-And even on Windows:
+or even easier:
+
+.. code:: bash
+
+    ./scripts/build.sh --Release #specifies the build type
```
Comment (chriseth) [inline docs/installing-solidity.rst]: I think it should be just `Release` but in general, I guess it is better not to confuse the readers here and just mention the script instead.

File: scripts/build.sh [hunk 6]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+
+if [ -z "$1" ]; then 
+    BUILD_TYPE=Release
+else 
+    BUILD_TYPE="$1"
+fi
+
+PWD=$(pwd)
```
Comment (chriseth) [inline scripts/build.sh]: Please do not assume a specific working directory, instead use `$(dirname $0)` to determine the location of the source files.

**Review by chriseth — CHANGES_REQUESTED** [review]

File: scripts/build.sh (hunk 6 above)
Comment (VoR0220) [reply scripts/build.sh]: See below, I use a git command to get us to the top of the root directory, this is merely so that we take note of wherever the user is executing from and return them to that spot when they are done.

File: scripts/Dockerfile (hunk 4 above)
Comment (VoR0220) [reply scripts/Dockerfile]: *sigh* alright...worth a shot.

File: docs/installing-solidity.rst (hunk 5 above)
Comment (VoR0220) [reply docs/installing-solidity.rst]: fair enough. This is all in the previous PR FYI...but there have been some problems...it seems I may have been too quick to be super confident that the rpc tests would pass like that.

File: .travis.yml (hunk 2 above)
Comment (VoR0220) [reply .travis.yml]: I created a template from the travis docs site, and yea I definitely agree looking at it now they ought to be switched.

File: .travis.yml (hunk 3 above)
Comment (VoR0220) [reply .travis.yml]: That's not a bad idea...but we have to be very strict about this.

File: .travis.yml (hunk 1 above)
Comment (VoR0220) [reply .travis.yml]: I think that's a little too open ended...would need to have something that checks for semver formatting

Comment (VoR0220) [reply .travis.yml]: was already planning on the script just wanted to see if it would pass.

Comment (chriseth) [reply .travis.yml]: Oh right, this is a bit complicated. We want the docker image to be created only for the tag, but only if it is run on the release branch (by the way, it is "release" and not "master"). TRAVIS_BRANCH will be set to the tag name if it runs on a tag. I think the following should do: If TRAVIS_BRANCH starts with v and the prerelease.txt file is empty.

File: .travis.yml (hunk 3 above)
Comment (chriseth) [reply .travis.yml]: You are right, probably better to grep for PROJECT_VERSION

File: scripts/build.sh (hunk 6 above)
Comment (chriseth) [reply scripts/build.sh]: ah ok. To restore the pwd, I usually just start a subshell:
```
(
do you stuff
)
```
bash will automatically take you back.

Comment (VoR0220) [reply scripts/build.sh]: It does? I didn't know that.

**Comment by VoR0220** [comment]
@chriseth looks like everything is passing properly. One last review of the docker_deploy script?

File: .travis.yml [hunk 7]
```diff
@@ -186,8 +186,11 @@ deploy:
     # Both the build and deploy steps for Emscripten are only run within the Ubuntu
     # configurations (not for macOS).  That is controlled by conditionals within the bash
     # scripts because TravisCI doesn't provide much in the way of conditional logic.
+    # This is also the deploy target for the dockerfile. If we are pushing into a develop branch, it will be tagged
+    # as a nightly and appended the commit of the branch it was pushed in. If we are pushing to master it will
+    # be tagged as "stable" and given the version tag as well.
     - provider: script
-      script: test $TRAVIS_EMSCRIPTEN != On || scripts/release_emscripten.sh
+      script: test $SOLC_EMSCRIPTEN == On && scripts/release_emscripten.sh || test $SOLC_DOCKER != On || ./scripts/docker_deploy.sh
```
Comment (chriseth) [inline .travis.yml]: I don't think it will work that way. If the emscripten release succeeded, it will not do the docker release.

File: scripts/Dockerfile [hunk 8]
```diff
@@ -1,12 +1,16 @@
...
-  apk del sed build-base git make cmake gcc g++ musl-dev curl-dev boost-dev                                 && \
-  rm -rf /var/cache/apk/*
+#Establish working directory as solidity
+WORKDIR /solidity
+#Copy working directory on travis to the image
+COPY / $WORKDIR
+
+#Install dependencies, eliminate annoying warnings, and build release, delete all remaining points and statically link.
+RUN ./scripts/install_deps.sh && sed -i -E -e 's/include <sys\/poll.h>/include <poll.h>/' /usr/include/boost/asio/detail/socket_types.hpp	&&\
+cmake -DCMAKE_BUILD_TYPE=Release -DTESTS=0 -DSTATIC_LINKING=1																				&&\
```
Comment (chriseth) [inline scripts/Dockerfile]: Please don't use tabs here.

File: scripts/Dockerfile [hunk 9]
```diff
@@ -1,12 +1,16 @@
...
-  git clone --depth 1 --recursive -b develop https://github.com/ethereum/solidity                           && \
-  cd /solidity && cmake -DCMAKE_BUILD_TYPE=Release -DTESTS=0 -DSTATIC_LINKING=1                             && \
-  cd /solidity && make solc && install -s  solc/solc /usr/bin                                               && \
-  cd / && rm -rf solidity                                                                                   && \
-  apk del sed build-base git make cmake gcc g++ musl-dev curl-dev boost-dev                                 && \
-  rm -rf /var/cache/apk/*
+#Establish working directory as solidity
+WORKDIR /solidity
+#Copy working directory on travis to the image
+COPY / $WORKDIR
```
Comment (chriseth) [inline scripts/Dockerfile:8]: Is `/` really the right directory?

Comment (chriseth) [reply scripts/Dockerfile:8]: Also, if we do it like that, it will not work as a standalone docker file anymore. Not sure what is "best practice" for docker files...

File: scripts/build.sh [hunk 10]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
+
+if [ -z "$1" ]; then 
+    BUILD_TYPE=Release
+else 
+    BUILD_TYPE="$1"
+fi
+
+cd $(git rev-parse --show-toplevel)
```
Comment (chriseth) [inline scripts/build.sh]: Please use things like `$(dirname "$0")` - we cannot assume that this is a git repository, for example the source tarballs don't have one.

File: scripts/docker_deploy.sh [hunk 11]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env sh
+
+docker login -u="$DOCKER_USERNAME" -p="$DOCKER_PASSWORD";
+if [ "$TRAVIS_BRANCH" == "develop" ]; then
+	version=`grep -oP "PROJECT_VERSION \"?\K[0-9.]+(?=\")"? CMakeLists.txt`;
+	docker tag ethereum/solc:build ethereum/solc:nightly;
+	docker tag ethereum/solc:build ethereum/solc:"$version"-"$TRAVIS_COMMIT"
+	docker push ethereum/solc:"$version"-"$TRAVIS_COMMIT";
+	docker push ethereum/solc:nightly;
+elif [ "$TRAVIS_PULL_REQUEST_BRANCH" == "master"]; then
```
Comment (chriseth) [inline scripts/docker_deploy.sh]: If we are currently building a pull request, we should never publish anything anywhere.

File: scripts/install_deps.sh [hunk 12]
```diff
@@ -94,19 +94,20 @@ case $(uname -s) in
...
-
-        # We should really 'brew install' our eth client here, but at the time of writing
-        # the bottle is known broken, so we will just cheat and use a hardcoded ZIP for
-        # the time being, which is good enough.   The cause of the breaks will go away
-        # when we commit the repository reorg changes anyway.
-        curl -L -O https://github.com/bobsummerwill/cpp-ethereum/releases/download/v1.3.0/cpp-ethereum-osx-mavericks-v1.3.0.zip
-        unzip cpp-ethereum-osx-mavericks-v1.3.0.zip
-
+        if ["$CI" = true]; then
+            brew update
```
Comment (chriseth) [inline scripts/install_deps.sh]: Please extract the common parts.

File: scripts/tests.sh [hunk 13]
```diff
@@ -54,7 +48,7 @@ $ETH_PATH --test -d /tmp/test &
 # is available and is ready for the unit-tests to start talking to it.
 while [ ! -S /tmp/test/geth.ipc ]; do sleep 2; done
 echo "--> IPC available."
-
+sleep 2
```
Comment (chriseth) [inline scripts/tests.sh:61]: Do you think this is needed?

File: scripts/Dockerfile (hunk 9 above)
Comment (VoR0220) [reply scripts/Dockerfile:8]: The dockerfile itself is not important to the user. What is important is the image that it produces. This enables us to easily automate the creation of an image.

File: .travis.yml (hunk 7 above)
Comment (VoR0220) [reply .travis.yml]: :/ was worried about that.

File: scripts/Dockerfile (hunk 9 above)
Comment (VoR0220) [reply scripts/Dockerfile:8]: Also yes. '/' is the right directory. Will find links but it basically has to do with the commands entry point.

File: scripts/build.sh (hunk 10 above)
Comment (VoR0220) [reply scripts/build.sh]: ++

File: scripts/docker_deploy.sh (hunk 11 above)
Comment (VoR0220) [reply scripts/docker_deploy.sh]: I figured this environmental var would persist even after the build succeeded and reached deploy.

File: scripts/tests.sh (hunk 13 above)
Comment (VoR0220) [reply scripts/tests.sh:61]: It helps with ensuring ipc channel is available.

File: scripts/install_deps.sh (hunk 12 above)
Comment (VoR0220) [reply scripts/install_deps.sh]: Good point

File: scripts/Dockerfile (hunk 9 above)
Comment (VoR0220) [reply scripts/Dockerfile:8]: furthermore, once the image is created, then the user can easily use it standalone by just doing `docker pull ethereum/solc:OurVersionTagHere`

Comment (VoR0220) [reply scripts/Dockerfile:8]: alternatively as opposed to COPY we could use a VOLUME of the current repo, but I figured that would be a less elegant and harder to understand command for anyone looking to improve upon this.

File: scripts/docker_deploy.sh (hunk 11 above)
Comment (VoR0220) [reply scripts/docker_deploy.sh]: ah...I see where my misunderstanding stemmed from. Okay. Yep. Changing.

File: .travis.yml (hunk 7 above)
Comment (VoR0220) [reply .travis.yml]: huh...appears they just changed it to allow multiple deploy scripts. I'll take full advantage of that :)

File: scripts/Dockerfile (hunk 9 above)
Comment (chriseth) [reply scripts/Dockerfile:8]: Ok, right. As long as this also works outside of travis automation with a normal  source checkout, I'm fine.

File: scripts/tests.sh (hunk 13 above)
Comment (chriseth) [reply scripts/tests.sh:61]: No, checking whether the file is available helps, that is done in line 59. What I was asking about was: Did you experience things like race conditions here that forced you to add the two second delay?

Comment (axic) [reply scripts/tests.sh:61]: Shouldn't the loop above ensure it?

Comment (VoR0220) [reply scripts/tests.sh:61]: @chriseth i thought maybe i was earlier on. Not sure if that was due to eliminating the if conditional on the eth branch. If you want to try restarting the build a number of times to see if it'll pass im okay with that.

File: scripts/Dockerfile (hunk 9 above)
Comment (VoR0220) [reply scripts/Dockerfile:8]: yes, it will work as long as you execute that build command in the travis.yml at the root of the repo.

File: scripts/tests.sh (hunk 13 above)
Comment (VoR0220) [reply scripts/tests.sh:61]: ^yup...see latest test....race condition. Will add it back in.

File: scripts/docker_deploy.sh [hunk 14]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env sh
+
+docker login -u="$DOCKER_USERNAME" -p="$DOCKER_PASSWORD";
+version="0.4.10"
```
Comment (axic) [inline scripts/docker_deploy.sh]: The version shouldn't be hardcoded here. Check how the other scripts determine the version.

**Review by axic — CHANGES_REQUESTED** [review]

File: scripts/docker_deploy.sh (hunk 14 above)
Comment (VoR0220) [reply scripts/docker_deploy.sh]: there are parts that are hardcoded but I understand your point. Maybe this upcoming commit will work?

**Comment by axic** [comment]
@VoR0220 reviewing this is a bit beyond me (never used docker before), but splitting out the regular build script is a PR I could review / approve. That is kind of separate to the other changes too.

**Comment by VoR0220** [comment]
Indeed I could probably split that out and get that approved...along with maybe the other travis.yml bits?

**Comment by axic** [comment]
I think the build script (+ the changelog and documentation bit) is good on its own. Travis could be separate to docker if you think that makes sense.

**Comment by VoR0220** [comment]
fair enough. Let me cut it out and push with a fresh branch.

File: scripts/Dockerfile [hunk 15]
```diff
@@ -1,12 +1,16 @@
 FROM alpine
 MAINTAINER chriseth <chris@ethereum.org>
+#Official solidity docker image
 
-RUN \
```
Comment (axic) [inline scripts/Dockerfile:4]: Can we keep this layout of commands? Changes are easier to see as well.

File: scripts/docker_deploy.sh [hunk 16]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env sh
+
+docker login -u="$DOCKER_USERNAME" -p="$DOCKER_PASSWORD";
+version=version=$(grep -oP "PROJECT_VERSION \"?\K[0-9.]+(?=\")"? $(dirname "$0")/CMakeLists.txt)
```
Comment (axic) [inline scripts/docker_deploy.sh]: Why `version=version=$(...`?

File: scripts/docker_deploy.sh [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+	docker tag ethereum/solc:build ethereum/solc:nightly-"$version"-"$TRAVIS_COMMIT"
+	docker push ethereum/solc:nightly-"$version"-"$TRAVIS_COMMIT";
+	docker push ethereum/solc:nightly;
+elif [ "$TRAVIS_BRANCH" == "release"]; then
+	docker tag ethereum/solc:build ethereum/solc:stable;
+	docker tag ethereum/solc:build ethereum/solc:"$version";
+	docker tag ethereum/solc:build ethereum/solc:
+	docker push ethereum/solc;
+	docker push ethereum/solc:"$version";
+	docker push ethereum/solc:stable;
```
Comment (axic) [inline scripts/docker_deploy.sh]: Is that `stable` or `release`?

File: scripts/install_deps.sh [hunk 18]
```diff
@@ -57,7 +57,7 @@ detect_linux_distro() {
         # extract 'foo' from NAME=foo, only on the line with NAME=foo
         DISTRO=$(sed -n -e 's/^NAME="\(.*\)\"/\1/p' /etc/os-release)
     elif [ -f /etc/centos-release ]; then
-	DISTRO=CentOS
+    DISTRO=CentOS
```
Comment (axic) [inline scripts/install_deps.sh]: Whitespace change.

File: scripts/Dockerfile (hunk 15 above)
Comment (VoR0220) [reply scripts/Dockerfile:4]: Not exactly. Plus tabs/spaces bit....

File: scripts/docker_deploy.sh (hunk 16 above)
Comment (VoR0220) [reply scripts/docker_deploy.sh]: good point. I guess I went too fast here.

File: scripts/docker_deploy.sh (hunk 17 above)
Comment (VoR0220) [reply scripts/docker_deploy.sh]: stable.

Comment (VoR0220) [reply scripts/docker_deploy.sh]: meaning the latest stable docker image.

File: scripts/docker_deploy.sh [hunk 19]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env sh
+
+docker login -u="$DOCKER_USERNAME" -p="$DOCKER_PASSWORD";
```
Comment (chriseth) [inline scripts/docker_deploy.sh:3]: This still has to be configured, right?

Comment (VoR0220) [reply scripts/docker_deploy.sh:3]: Configured in Travis, yes.

File: scripts/docker_deploy.sh [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+version=$(grep -oP "PROJECT_VERSION \"?\K[0-9.]+(?=\")"? $(dirname "$0")/CMakeLists.txt)
+if [ "$TRAVIS_BRANCH" == "develop" ]; then
+	docker tag ethereum/solc:build ethereum/solc:nightly;
+	docker tag ethereum/solc:build ethereum/solc:nightly-"$version"-"$TRAVIS_COMMIT"
+	docker push ethereum/solc:nightly-"$version"-"$TRAVIS_COMMIT";
+	docker push ethereum/solc:nightly;
+elif [ "$TRAVIS_BRANCH" == "release"]; then
```
Comment (chriseth) [inline scripts/docker_deploy.sh:10]: It should not check for the branch, it should check for the `v...` tag on that branch.

Comment (VoR0220) [reply scripts/docker_deploy.sh:10]: How do you write that conditional in an automated way?

**Comment by VoR0220** [comment]
@chriseth ready to merge whenever you're ready.

File: scripts/docker_deploy.sh (hunk 20 above)
Comment (VoR0220) [reply scripts/docker_deploy.sh:10]: done, see below.

**Review by chriseth — APPROVED** [review]
